### PR TITLE
Bugfix for permission management of entities

### PR DESCRIPTION
**Pull request by crazyscientist**
A small bug allowed authenticated users to perform any actions on entities without permission, when the `reviewable` flag of the schema was unset. On the flip side, it prevented them from creating change requests for entities they had no permissions on.

This change fixes the situation:

* Any authenticated user is allowed to create change requests
* Only authorized users are allowed to apply changes to entities

### CODE REVIEW

File: backend/tests/conftest.py [hunk 1]
```diff
@@ -353,6 +353,22 @@ def delete(self, url, *, params=None, headers=None, cookies=None, auth=USE_CLIEN
...
+
+
+@pytest.fixture
+def unauthorized_testuser(dbsession, testuser) -> User:
+    # The testuser is a superuser. Remove permissions.
+    dbsession.query(Permission).filter(Permission.recipient_type == RecipientType.USER,
+                                       Permission.recipient_id == testuser.id).delete()
+    return testuser
+
+
```
Comment (der-gabe) [inline backend/tests/conftest.py:371]: Why is this here twice, i.e. doubled?

Looks identical to me…

Oversight, or am I missing something?

File: backend/tests/test_auth_api.py [hunk 2]
```diff
@@ -385,12 +385,61 @@ def test_require_authentication(self, method, url, client: TestClient):
...
+            .join(Schema)\
+            .filter(Schema.slug == "unperson", Entity.slug == "foo")\
+            .exists()
+        assert not dbsession.query(Entity.id).filter(q).scalar()
+
+    def test_review_change_request_breakout__update(self, dbsession: Session,
+                                                    unauthorized_testuser: User,
+                                                    authenticated_client: TestClient):
+        """
+        Test that authenticated user is allowed to create a change request without actually changing
```
Comment (der-gabe) [inline backend/tests/test_auth_api.py:414]:
```suggestion
        Test that authenticated user is allowed to update a change request without actually changing
```

File: backend/tests/test_auth_api.py [hunk 3]
```diff
@@ -385,12 +385,61 @@ def test_require_authentication(self, method, url, client: TestClient):
...
+
+        assert response.status_code == 202
+        dbsession.refresh(entity)
+        assert entity.name == "Foo"
+
+    def test_review_change_request_breakout__delete(self, dbsession: Session,
+                                                    unauthorized_testuser: User,
+                                                    authenticated_client: TestClient):
+        """
+        Test that authenticated user is allowed to create a change request without actually deleting
```
Comment (der-gabe) [inline backend/tests/test_auth_api.py:432]:
```suggestion
        Test that authenticated user is allowed to delete a change request without actually deleting
```

File: backend/tests/conftest.py (hunk 1 above)
Comment (crazyscientist) [reply backend/tests/conftest.py:371]: It should suffice to say: Error between keyboard and chair. Let me delete the duplication...

File: backend/tests/test_auth_api.py [hunk 4]
```diff
@@ -385,12 +385,61 @@ def test_require_authentication(self, method, url, client: TestClient):
...
+        an entity
+        """
+        schema = dbsession.query(Schema).where(Schema.slug == "unperson").one()
+        entity = Entity(name="Foo", slug="foo", schema_id=schema.id)
+        dbsession.add(entity)
+        dbsession.commit()
+        response = authenticated_client.delete("/entity/unperson/foo",
+                                               json={"name": "Føø", "slug": "foo"})
```
Comment (der-gabe) [inline backend/tests/test_auth_api.py:440]: Do we even need the `json` argument in this case?

And if so, shouldn't it match the original entity's values?

File: backend/tests/test_auth_api.py [hunk 5]
```diff
@@ -385,12 +385,61 @@ def test_require_authentication(self, method, url, client: TestClient):
...
 
     @pytest.mark.parametrize("method, url", authorized)
-    def test_require_authorization(self, method, url, dbsession: Session, testuser: User,
-                                   authenticated_client: TestClient):
-        # The testuser is a superuser. Remove permissions first.
-        dbsession.query(Permission).filter(Permission.recipient_type == RecipientType.USER,
-                                           Permission.recipient_id == testuser.id).delete()
-
+    def test_require_authorization(self, method, url, dbsession: Session,
+                                   unauthorized_testuser: User, authenticated_client: TestClient):
```
Comment (der-gabe) [inline backend/tests/test_auth_api.py:389]: This test has only been refactored, bu it didn't catch the original bug before, so wouldn't one or a few new tests be in order?

File: backend/tests/test_auth_api.py (hunk 4 above)
Comment (crazyscientist) [reply backend/tests/test_auth_api.py:440]: Right, that's what you get for working late in the evening...

```suggestion
        response = authenticated_client.delete("/entity/unperson/foo")
```

File: backend/tests/test_auth_api.py (hunk 5 above)
Comment (crazyscientist) [reply backend/tests/test_auth_api.py:389]: Yes, that's what the new `test_review_change_request_breakout_*` are for :slightly_smiling_face: 

We don't want an unauthorized user to change any entity data, but they should be allowed to submit a request

Comment (crazyscientist) [reply backend/tests/test_auth_api.py:389]: But we have not covered the case of an unauthorized user accepting the request. Let me add that...

File: backend/tests/test_auth_api.py [hunk 6]
```diff
@@ -385,12 +387,72 @@ def test_require_authentication(self, method, url, client: TestClient):
...
         func = getattr(authenticated_client, method)
         response = func(url)
         assert response.status_code == 403
+
+    def test_approve_request(self, dbsession: Session, unauthorized_testuser: User,
+                             authenticated_client: TestClient):
+        change_request_ids = dbsession.query(ChangeRequest.id)\
```
Comment (der-gabe) [inline backend/tests/test_auth_api.py:398]:
```suggestion
        """
        Test that an authenticated but unauthorized user is not allowed to approve a change request.
        """
        change_request_ids = dbsession.query(ChangeRequest.id)\
```

File: backend/tests/test_auth_api.py (hunk 5 above)
Comment (der-gabe) [reply backend/tests/test_auth_api.py:391]: That's great, but my point was that apparently, if I understand the issue correctly, authenticated, but unauthorized users were able to make any kind of change they wanted – directly, i.e. w/o a change request.

Is that correct?

Because if so, then why wasn't *that* caught by any test?

The new tests merely check that such users can create change _requests_ (which themselves don't change anything directly), but isn't there also a way to make changes w/o requests? And if so, where is there a test that unauthorized users cannot make such changes?

Comment (crazyscientist) [reply backend/tests/test_auth_api.py:391]: > why wasn't that caught by any test?

Because the existing test (intended to verify permissions on direct entity changed) checked against the wrong schema (i.e. the one which required reviews). So, the test incorrectly checked that unauthorized users were not allowed to create a change request. By changing to the schema with an unset `reviewable` flag, the test checks what is actually intended: Unauthorized users are not allowed to change an entity.

The new tests cover the other way of doing things: Unauthorized users may create change requests, when the `reviewable` flag is set, but not apply them to the entity.

Does that clarify things?

Comment (der-gabe) [reply backend/tests/test_auth_api.py:391]: > Because the existing test […]

i.e. `test_require_authorization`, right?

> […] checked against the wrong schema (i.e. the one which required reviews). So, the test incorrectly checked that unauthorized users were not allowed to create a change request. By changing to the schema with an unset `reviewable` flag, the test checks what is actually intended: Unauthorized users are not allowed to change an entity.

This implies that the existing **test** should have been _functionally_ changed in some way (to test the right schema), but it _hasn't!_ The only thing that has changed is that some setup code has been moved into a fixture. Great, but that's only a refactoring.

The test used to pass, when the bug was present, and it hasn't functionally changed.

Why should I now have confidence in a test that previously did not fail in the presence of the exact kind of bug it's meant to detect?

If I were to reintroduce the same bug again, would `test_require_authorization` now fail?

If so, I don't understand why.

If not, then there's still at least one test missing.

I'm open to having a call about this, but ATM it's super unclear.

> The new tests cover the other way of doing things: […]

That part is clear to me and well done for adding those tests.

Comment (crazyscientist) [reply backend/tests/test_auth_api.py:391]: Yes, we are talking about `test_require_authorization`, which is also a parameterized test. In the parameter definition the schema was changed from `unperson` to `person`.

If you were to reintroduce the bug, the testsuite would fail as expected.

If you like, give me a ring...

Comment (der-gabe) [reply backend/tests/test_auth_api.py:391]: Ah... :bulb: the parametrisation changed!

I meant to ask about that change from `unperson` to `person` and how it ties in, but then I forgot.

Now I see it and it makes sense.

**Review by der-gabe — APPROVED** [review]

**Comment by crazyscientist** [comment]
Thank you :slightly_smiling_face:
